kafka/conn.py

```diff
@@ -808,24 +819,26 @@ def send_pending_requests(self):
     def can_send_more(self):
         """Return True unless there are max_in_flight_requests_per_connection."""
         max_ifrs = self.config['max_in_flight_requests_per_connection']
-        return len(self.in_flight_requests) < max_ifrs
+        with self._ifr_lock:
+            return len(self.in_flight_requests) < max_ifrs
 
     def recv(self):
         """Non-blocking network receive.
 
         Return list of (response, future) tuples
         """
-        if not self.connected() and not self.state is ConnectionStates.AUTHENTICATING:
-            log.warning('%s cannot recv: socket not connected', self)
-            # If requests are pending, we should close the socket and
-            # fail all the pending request futures
-            if self.in_flight_requests:
-                self.close(Errors.KafkaConnectionError('Socket not connected during recv with in-flight-requests'))
-            return ()
+        with self._ifr_lock:
+            if not self.connected() and not self.state is ConnectionStates.AUTHENTICATING:
+                log.warning('%s cannot recv: socket not connected', self)
+                # If requests are pending, we should close the socket and
+                # fail all the pending request futures
+                if self.in_flight_requests:
+                    self.close(Errors.KafkaConnectionError('Socket not connected during recv with in-flight-requests'))
```

Comment:
`close` also acquires the ifr_lock, so this is going to fail unless the lock is reentrant

Reply:
Ah, not another one :(
Thanks!